File: composeApp/src/androidMain/java/org/application/shikiapp/shared/CaptionStyleCompat.java
package org.application.shikiapp.shared;

import static java.lang.annotation.ElementType.TYPE_USE;

import android.graphics.Color;
import android.graphics.Typeface;
import android.view.accessibility.CaptioningManager;

import androidx.annotation.IntDef;
import androidx.annotation.Nullable;
import androidx.media3.common.util.UnstableApi;

import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

@UnstableApi
public record CaptionStyleCompat(
        int foregroundColor,
        int backgroundColor,
        int windowColor,
        @EdgeType int edgeType, int edgeColor,
        @Nullable Typeface typeface
) {

    @Documented
    @Retention(RetentionPolicy.SOURCE)
    @Target(TYPE_USE)
    @IntDef({EDGE_TYPE_NONE, EDGE_TYPE_OUTLINE, EDGE_TYPE_DROP_SHADOW, EDGE_TYPE_RAISED, EDGE_TYPE_DEPRESSED})
    public @interface EdgeType {
    }

    public static final int EDGE_TYPE_NONE = 0;

    public static final int EDGE_TYPE_OUTLINE = 1;

    public static final int EDGE_TYPE_DROP_SHADOW = 2;

    public static final int EDGE_TYPE_RAISED = 3;

    public static final int EDGE_TYPE_DEPRESSED = 4;

    public static final int USE_TRACK_COLOR_SETTINGS = 1;

    public static final CaptionStyleCompat DEFAULT = new CaptionStyleCompat(
            Color.WHITE,
            Color.BLACK,
            Color.TRANSPARENT,
            EDGE_TYPE_NONE,
            Color.WHITE,
            null
    );

    public static CaptionStyleCompat createFromCaptionStyle(
            CaptioningManager.CaptionStyle captionStyle) {
        return new CaptionStyleCompat(
                captionStyle.hasForegroundColor() ? captionStyle.foregroundColor : DEFAULT.foregroundColor,
                captionStyle.hasBackgroundColor() ? captionStyle.backgroundColor : DEFAULT.backgroundColor,
                captionStyle.hasWindowColor() ? captionStyle.windowColor : DEFAULT.windowColor,
                captionStyle.hasEdgeType() ? captionStyle.edgeType : DEFAULT.edgeType,
                captionStyle.hasEdgeColor() ? captionStyle.edgeColor : DEFAULT.edgeColor,
                captionStyle.getTypeface());
    }
}

File: composeApp/src/androidMain/kotlin/org/application/shikiapp/shared/utils/ui/subtitles/CaptionStyleCompat.kt
package org.application.shikiapp.shared.utils.ui.subtitles

import android.graphics.Color
import android.graphics.Typeface
import android.view.accessibility.CaptioningManager
import androidx.annotation.IntDef
import androidx.media3.common.util.UnstableApi

@UnstableApi
data class CaptionStyleCompat(
    val foregroundColor: Int,
    val backgroundColor: Int,
    val windowColor: Int,
    @EdgeType val edgeType: Int,
    val edgeColor: Int,
    val typeface: Typeface?
) {

    @MustBeDocumented
    @Retention(AnnotationRetention.SOURCE)
    @Target(AnnotationTarget.TYPE, AnnotationTarget.VALUE_PARAMETER, AnnotationTarget.PROPERTY)
    @IntDef(
        EDGE_TYPE_NONE,
        EDGE_TYPE_OUTLINE,
        EDGE_TYPE_DROP_SHADOW,
        EDGE_TYPE_RAISED,
        EDGE_TYPE_DEPRESSED
    )
    annotation class EdgeType

    companion object {
        const val EDGE_TYPE_NONE = 0
        const val EDGE_TYPE_OUTLINE = 1
        const val EDGE_TYPE_DROP_SHADOW = 2
        const val EDGE_TYPE_RAISED = 3
        const val EDGE_TYPE_DEPRESSED = 4

        const val USE_TRACK_COLOR_SETTINGS = 1

        val DEFAULT = CaptionStyleCompat(
            foregroundColor = Color.WHITE,
            backgroundColor = Color.BLACK,
            windowColor = Color.TRANSPARENT,
            edgeType = EDGE_TYPE_NONE,
            edgeColor = Color.WHITE,
            typeface = null
        )

        fun createFromCaptionStyle(captionStyle: CaptioningManager.CaptionStyle) =
            CaptionStyleCompat(
                foregroundColor = if (captionStyle.hasForegroundColor()) captionStyle.foregroundColor else DEFAULT.foregroundColor,
                backgroundColor = if (captionStyle.hasBackgroundColor()) captionStyle.backgroundColor else DEFAULT.backgroundColor,
                windowColor = if (captionStyle.hasWindowColor()) captionStyle.windowColor else DEFAULT.windowColor,
                edgeType = if (captionStyle.hasEdgeType()) captionStyle.edgeType else DEFAULT.edgeType,
                edgeColor = if (captionStyle.hasEdgeColor()) captionStyle.edgeColor else DEFAULT.edgeColor,
                typeface = captionStyle.typeface
            )
    }
}